math/int.go

```diff
@@ -420,10 +421,90 @@ func (i *Int) Unmarshal(data []byte) error {
 	return nil
 }
 
+var bigMaxUint64 = new(big.Int).SetUint64(uint64(stdmath.MaxUint64))
+
 // Size implements the gogo proto custom type interface.
-func (i *Int) Size() int {
-	bz, _ := i.Marshal()
-	return len(bz)
+// Reduction power of 10 is the smallest power of 10, than 1<<64-1
+//
+//	18446744073709551615
+//
+// and the next value fitting with the digits of (1<<64)-1 is:
+//
+//	10000000000000000000
+var big10Pow19, _ = new(big.Int).SetString("1"+strings.Repeat("0", 19), 10)
+var log10Of2 = stdmath.Log10(2)
+
+func (i *Int) Size() (size int) {
+	sign := i.Sign()
+	if sign == 0 { // It is zero.
+		// log*(0) is undefined hence return early.
+		return 1
+	}
+
+	ii := i.i
+	alreadyMadeCopy := false
+	if sign < 0 { // Negative sign encountered, so consider len("-")
+		// The reason that we make this comparison in here is to
+		// allow checking for negatives exactly once, to reduce
+		// on comparisons inside sizeBigInt, hence we make a copy
+		// of ii and make it absolute having taken note of the sign
+		// already.
+		size += 1
+		// We already accounted for the negative sign above, thus
+		// we can now compute the length of the absolute value.
+		ii = new(big.Int).Abs(ii)
+		alreadyMadeCopy = true
+	}
+
+	// From here on, we are now dealing with non-0, non-negative values.
+	return size + sizeBigInt(ii, alreadyMadeCopy)
+}
+
+func sizeBigInt(i *big.Int, alreadyMadeCopy bool) (size int) {
+	// This code assumes that non-0, non-negative values have been passed in.
+	bitLen := i.BitLen()
+	res := float64(bitLen) * log10Of2
+	ires := int(res)
+	if diff := res - float64(ires); diff == 0.0 {
+		return size + ires
+	} else if diff >= 0.3 { // There are other digits past the bitLen, this is a heuristic.
+		return size + ires + 1
+	}
+
+	// Use Log10(x) for values less than (1<<64)-1, given it is only defined for [1, (1<<64)-1]
+	if i.Cmp(bigMaxUint64) <= 0 {
```

Comment:
isn;t this equivalent to bitlen < 64?

Reply:
It is and the result in changing it to `bitLen <= 64` isn't significant! I used bigMaxUint64 to make for clear readability and because when writing out the algorithm I was thinking in terms of log2 where a value like: 18446744073709551616, that value is +1 greater than maxUint64. If you pass in `math.Log10(float64(i.Uint64()))` would result in -Inf. Sure I'll change it to just that comparison